### DISCUSSION TITLE
[RFC FS-1099] Compiler helpers for list mutators

### DISCUSSION THREAD

**Discussion by dsyme** [fsharp/fslang-design#565] [Language and core library RFC discussions]
This is a discussion thread for https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1099-list-collector.md

This adds to FSharp.Core intrinsics to do tail-cons mutation specifically designed for fast creation of F# lists by a `list { ... }` builder.  This is useful from the outside and I'm 100% certain this will be needed for the inlined code for faster implementations of `[ ... ]` in the future.  So I think we should just add it now (it is in https://github.com/dotnet/fsharp/pull/6811)

**Comment by krauthaufen** [comment]
Hey, I'd be really happy to see this, I proposed something very similar with a "safer" API here: https://github.com/fsharp/fslang-suggestions/issues/926

I think this level of safety around the actual mutation would not cost much and make the whole thing easier to handle. Let me know what you think.

**Reply by krauthaufen** [reply]
Real implementation (outside Fsharp.Core, hence using reflection to set the tail) is here: https://github.com/krauthaufen/DawnSharp/blob/master/src/Armadillo/Tools/ListBuilder.fs#L25

**Reply by dsyme** [reply]
Thanks, I've linked your suggestion in the RFC

**Comment by nojaf** [comment]
Out of interest, why does the instance of a ListCollector need to be `mutable`?
Leaving that out won't add any items when `.Add()` is being called.

**Reply by krauthaufen** [reply]
I also stumbled over that, it's simply because it is a struct i think and the first element being added mutates the struct itself (it creates a `Cons` node and stores it in the struct's mutable field). So when struct is copied before calling the add member you will not see the change outside. If my theory is correct this should only be necessary for the first element and all subsequent elements should also work if the `ListCollector` itself is not mutable.
